libs/@local/graph/types/typescript/src/validation.ts

```diff
@@ -0,0 +1,235 @@
+  DataTypeInferenceError as DataTypeInferenceErrorGraphApi,
+  EntityTypesError as EntityTypesErrorGraphApi,
+  EntityValidationReport as EntityValidationReportGraphApi,
+  JsonSchemaValueTypeMismatch as JsonSchemaValueTypeMismatchGraphApi,
+  LinkDataStateError as LinkDataStateErrorGraphApi,
+  LinkedEntityError as LinkedEntityErrorGraphApi,
+  LinkError as LinkErrorGraphApi,
+  LinkTargetError as LinkTargetErrorGraphApi,
+  LinkValidationReport as LinkValidationReportGraphApi,
+  MetadataValidationReport as MetadataValidationReportGraphApi,
+  ObjectPropertyValidationReport as ObjectPropertyValidationReportGraphApi,
+  ObjectValidationReport as ObjectValidationReportGraphApi,
+  PropertyArrayValidationReport as PropertyArrayValidationReportGraphApi,
+  PropertyObjectValidationReport as PropertyObjectValidationReportGraphApi,
+  PropertyValidationReport as PropertyValidationReportGraphApi,
+  PropertyValueTypeMismatch as PropertyValueTypeMismatchGraphApi,
+  PropertyValueValidationReport as PropertyValueValidationReportGraphApi,
+  Report,
+  UnexpectedEntityType as UnexpectedEntityTypeGraphApi,
+  ValueValidationError as ValueValidationErrorGraphApi,
+  ValueValidationReport as ValueValidationReportGraphApi,
+} from "@local/hash-graph-client";
+
+export type EntityValidationReport = Omit<
+  EntityValidationReportGraphApi,
+  "properties" | "link" | "metadata"
+> & {
+  link?: LinkValidationReport;
+  metadata?: MetadataValidationReport;
+} & ObjectValidationReport;
+
+export type ObjectPropertyValidationReport = Subtype<
+  ObjectPropertyValidationReportGraphApi,
+  | {
+      // The property was found at the entity but is not expected in the schema
+      type: "unexpected";
+    }
+  | {
+      // It was not possible for the graph to read the property type
+      type: "retrieval";
+      error: Report;
+    }
+  | {
+      // The property was found at the entity but the type is not the expected type from Athe schema
+      type: "wrongType";
+      data: PropertyValueTypeMismatchGraphApi;
+    }
+  | PropertyValidationReport
+  | ({
+      // The validation of the property array failed
+      type: "propertyArray";
+    } & ArrayValidationReport)
+  | {
+      // The property is required by the schema but is missing from the entity
+      type: "missing";
+    }
+>;
+
+export type JsonSchemaValueTypeMismatch = Omit<
+  JsonSchemaValueTypeMismatchGraphApi,
+  "actual" | "expected"
+> & {
+  actual: [VersionedUrl, ...VersionedUrl[]];
+  expected: [VersionedUrl, ...VersionedUrl[]];
+};
+
+export type PropertyValidationReport = Subtype<
+  PropertyValidationReportGraphApi,
+  | {
+      // The property value validation failed
+      type: "value";
+      /*
+       * Validation for each constraint in the `oneOf` field. The validation is assumed to pass if exactly one of the
+       * constraints passes. In this case, this field will be omitted. Whenever this field is present it can be assumed
+       * that the validation failed.
```

Comment:
Exactly one, or at least one? We want the latter (and therefore want to rename `oneOf` to `anyOf` at some point)

Reply:
Currently it's exactly one. We can change this as part of this PR if you like.

Reply:
Let's change it now so it doesn't trip us up later

Reply:
https://github.com/hashintel/hash/pull/5850/files/8fa0baa5fa34141623ff8e99f5733f3ad388bf91..59f72104d3cc1f879c009218225a1cdd28adc8f6